test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,116 @@
+
+	// Removes all the finalizers.
+	// We are testing the worst-case scenario, i.e. all finalizers are deleted, even when reconciliation is paused.
+	// Deletion of the finalizers can either happen from a system generated attack or by mistake, which is an edge case.
+	// The reconciliation loop should be able to recover from this, by adding the required finalizers back.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
+
+	// Once all Clusters are paused remove the finalizers from all objects in the graph.
+	removeFinalizers(ctx, proxy, namespace)
+
+	// Unpause the cluster.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the clusterClass, if one is in use, to speed up reconciliation. This ensures ClusterClass finalizers
+	// are re-reconciled before asserting the owner reference graph.
+	forceClusterClassReconcile(ctx, proxy.GetClient(), clusterKey)
+
+	// Check that the finalizers are as expected after further reconciliations.
+	assertFinalizersExist(ctx, proxy, namespace, initialFinalizers)
+}
+
+func getFinalizersFromAllObjs(ctx context.Context, proxy ClusterProxy, namespace string) map[string][]string {
+	finalizers := map[string][]string{}
+
+	graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace, proxy.GetKubeconfigPath())
+	Expect(err).ToNot(HaveOccurred())
+
+	for _, v := range graph {
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion(v.Object.APIVersion)
+		obj.SetKind(v.Object.Kind)
+		err = proxy.GetClient().Get(ctx, client.ObjectKey{Namespace: v.Object.Namespace, Name: v.Object.Name}, obj)
+		Expect(err).ToNot(HaveOccurred())
+		if obj.GetFinalizers() != nil {
+			finalizers[fmt.Sprintf("%s/%s", v.Object.Kind, client.ObjectKey{Name: v.Object.Name, Namespace: v.Object.Namespace}.String())] = obj.GetFinalizers()
+		}
+	}
+
+	return finalizers
+}
+
+func removeFinalizers(ctx context.Context, proxy ClusterProxy, namespace string) {
+	graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace, proxy.GetKubeconfigPath())
+	Expect(err).ToNot(HaveOccurred())
+	for _, object := range graph {
+		ref := object.Object
+		obj := new(unstructured.Unstructured)
+		obj.SetAPIVersion(ref.APIVersion)
+		obj.SetKind(ref.Kind)
+		obj.SetName(ref.Name)
+
+		Expect(proxy.GetClient().Get(ctx, client.ObjectKey{Namespace: namespace, Name: object.Object.Name}, obj)).To(Succeed())
+		helper, err := patch.NewHelper(obj, proxy.GetClient())
+		Expect(err).ToNot(HaveOccurred())
+		obj.SetFinalizers([]string{})
+		Expect(helper.Patch(ctx, obj)).To(Succeed())
+	}
+}
+
+func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace string, initialFinalizers map[string][]string) {
+	Eventually(func() error {
+		var allErrs []error
+		afterDeleteFinalizers := getFinalizersFromAllObjs(ctx, proxy, namespace)
+
+		missing := map[string][]string{}
+		for k, v := range initialFinalizers {
+			if _, ok := afterDeleteFinalizers[k]; !ok {
```

Comment:
WDYT about making the list of finalizers explicit per type like we do with ownerReferences? This is just as good of a test for re-reconciliation, but adding explicit assert functions for the finalizers we expect per kind would help to document our expected state which I think is useful.

Reply:
Having looked at the owner reference test implementation, I was thinking about the same, but wasn't sure. Good idea, I will implement that.